pkg/cableengine/healthchecker/pinger.go

```diff
@@ -0,0 +1,138 @@
+package healthchecker
+
+import (
+	"math"
+	"time"
+
+	"github.com/go-ping/ping"
+	"k8s.io/klog"
+)
+
+var waitTime time.Duration = 15
+var size uint64 = 1000
+
+type Pinger struct {
+	remoteEndpointName string
+	healthCheckIp      string
+	statistics         Statistics
+	Status             string
+	stopCh             chan struct{}
+	pinger             *ping.Pinger
+}
+
+type Statistics struct {
+	previousRtts []uint64
+	sum          uint64
+	mean         uint64
+	stdDev       uint64
+	lastRtt      uint64
+	minRtt       uint64
+	maxRtt       uint64
+	sqrDiff      uint64
+	index        uint64
+	size         uint64
+}
+
+func NewPinger(localEndpointName, healthCheckIP string) *Pinger {
+	return &Pinger{
+		remoteEndpointName: localEndpointName,
+		healthCheckIp:      healthCheckIP,
+		statistics: Statistics{
+			size:         size,
+			previousRtts: make([]uint64, size),
+		},
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (s *Statistics) updateStatistics(rtt uint64) {
+	s.lastRtt = rtt
+
+	if (s.index + 1) > 1 {
+		s.previousRtts[s.index%s.size] = rtt
+		if s.minRtt == 0 || s.minRtt > rtt {
+			s.minRtt = rtt
+		}
+
+		if s.maxRtt < rtt {
+			s.maxRtt = rtt
+		}
+
+		s.sum += rtt
+		oldMean := s.mean
+		s.mean = s.sum / (s.index + 1)
+		s.sqrDiff += uint64(((int64(rtt - oldMean)) * int64((rtt - s.mean))))
+		s.stdDev = uint64(math.Sqrt(float64(s.sqrDiff / (s.index + 1))))
+	} else {
+		s.sum = rtt
+		s.sqrDiff = 0
+		s.minRtt = rtt
+		s.maxRtt = rtt
+		s.mean = rtt
+		s.sum = rtt
+		s.previousRtts[s.index%s.size] = rtt
+		s.stdDev = 0
+	}
+
+	if (s.index + 1) == s.size {
+		// Resetting since the incremental SD calculated have an error factor due to truncation which
+		// could be significant as count increases.
+		s.index = 1
+		s.previousRtts[0] = s.previousRtts[s.size-2]
+		s.previousRtts[1] = s.previousRtts[s.size-1]
+		s.sum = s.previousRtts[0] + s.previousRtts[1]
+		s.mean = s.sum / 2
```

Comment:
This will make the mean limited just to a few samples. 

I think we should probably update this implementation later to do a moving average (for example, for the last N hrs.) 


@nyechiel  ^ what do you think would be most useful? 1h / 2h/ 4h/ .. 24h?

Reply:
Can we add a TODO to handle this on the future?

Reply:
I will add a TODO.

We can do a moving average with this algorithm. But since there an approximation when we calculate the square root, an error factor will be introduced. With a 100 samples itself, I could see a small difference with this algorithm for SD ( which computes incremental SD for each element addition)  in the least significant digits. This will grow as the packet count increases.

So we will have to reset the SD periodically.  If we make window large say 24 hours, we will have some computational overhead to recalculate SD for that period.

Reply:
Added TODO.